src/mpid/ch4/netmod/ucx/ucx_rma.h

```diff
@@ -308,9 +308,11 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_NM_mpi_put(const void *origin_addr,
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_NM_MPI_PUT);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_NM_MPI_PUT);
 
-    if (!MPIDI_UCX_is_reachable_target(target_rank, win, winattr)) {
-        mpi_errno = MPIDIG_mpi_put(origin_addr, origin_count, origin_datatype, target_rank,
-                                   target_disp, target_count, target_datatype, win);
+    if (!MPIDI_UCX_is_reachable_target(target_rank, win, winattr) ||
+        MPL_gpu_query_pointer_is_dev(origin_addr)) {
+        mpi_errno =
+            MPIDIG_mpi_put(origin_addr, origin_count, origin_datatype, target_rank, target_disp,
+                           target_count, target_datatype, win);
     } else {
```

Comment:
Do we also need check the window buffer at window creation so that we can disable native path if it is GPU?

Reply:
That is a good idea.

Reply:
Hm, yeah I guess we could stash that in `winattr`.

Reply:
Can be add into the check for `MPIDI_WINATTR_NM_REACHABLE`

Reply:
Yes, I considered that at one point but didn't do it for some reason. I will revisit before posting the next version.

Reply:
Oh, we'll need to add a couple arguments to handle the buffer pointers. I guess that's not any worse than adding another function call.

Reply:
I wonder if `fi_mr_reg` will fail for device memory. In that case, this should already be handled by the existing window hooks. I will run a test to confirm.

Reply:
It depends on provider. If the provider doesn't do hardware registration, then most likely it doesn't check the origin of memory.

Reply:
> Oh, we'll need to add a couple arguments to handle the buffer pointers. I guess that's not any worse than adding another function call.

I don't get it. I guess we need make a fast path for cases where we use only host memory. For instance, we need check only a single bit to figure out whether the win buffer is device memory. If it eventually goes to device memory, then it is already in slow path and we don't care overheads from pointer accesses or function calls :-)

Reply:
Sorry, I was confusing 2 different things. The reachable attribute works for the target memory region. The `MPIDI_OFI|UCX_is_reachable_target` function used in the communication path also needs to check the pointers for origin, result, and compare addresses to make sure they are not also on the GPU.

Reply:
Got it. I was not aware that you add the origin side check also into `MPIDI_UCX_is_reachable_target`. Then it makes sense that we cannot get rid of the pointer access for checking origin side buffers, as long as the build enables cuda. Hopefully the pointers are not checked again inside UCX or OFI

Reply:
Can we reformat line 311 manually to avoid the code cleanup change? Same for the other RMA functions in `ucx_rma.h`

Reply:
Hmm, yeah that is weird. Let me see if I can fix it.

Reply:
Fixed.